op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -421,37 +454,54 @@ func runStateMigration(newDBPath string, opts stateMigrationOptions) error {
 
-	// Get latest block information from L1
 	var l1StartBlock *types.Block
 	client, err := ethclient.Dial(opts.l1RPC)
 	if err != nil {
 		return fmt.Errorf("cannot dial %s: %w", opts.l1RPC, err)
 	}
-
+	// If the L1 starting block tag is not set, we determine it dynamically by
+	// finding the most recent final L1 block at the time of the L2 fork block.
 	if config.L1StartingBlockTag == nil {
-		l1StartBlock, err = client.BlockByNumber(context.Background(), nil)
+		// Find the L1 starting block, the L2 fork block occurs 1 minute after the last celo L1 block.
+		opts.migrationBlockTime = celoL1Head.Time + 60
+		// Wait for the L2 start time before we calculate the L1 starting block.
+		tillL2Start := int64(opts.migrationBlockTime) - time.Now().Unix()
+		if tillL2Start >= 0 {
+			time.Sleep(time.Duration(tillL2Start+1) * time.Second)
+		}
+		l1StartingBlockHash, err := NewBeaconClient(opts.l1BeaconRPC, opts.l1BeaconchainURL).MostRecentFinalizedL1BlockAtTime(opts.migrationBlockTime)
 		if err != nil {
-			return fmt.Errorf("cannot fetch latest block: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock based on L2 starting block time time (%v): %w", opts.migrationBlockTime, err)
 		}
-		tag := rpc.BlockNumberOrHashWithHash(l1StartBlock.Hash(), true)
-		config.L1StartingBlockTag = (*genesis.MarshalableRPCBlockNumberOrHash)(&tag)
-	} else if config.L1StartingBlockTag.BlockHash != nil {
+		config.L1StartingBlockTag = &genesis.MarshalableRPCBlockNumberOrHash{BlockHash: &l1StartingBlockHash}
+	}
+
+	if config.L1StartingBlockTag.BlockHash != nil {
 		l1StartBlock, err = client.BlockByHash(context.Background(), *config.L1StartingBlockTag.BlockHash)
 		if err != nil {
-			return fmt.Errorf("cannot fetch block by hash: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock by hash (%v): %w", config.L1StartingBlockTag.BlockHash, err)
 		}
 	} else if config.L1StartingBlockTag.BlockNumber != nil {
 		l1StartBlock, err = client.BlockByNumber(context.Background(), big.NewInt(config.L1StartingBlockTag.BlockNumber.Int64()))
 		if err != nil {
-			return fmt.Errorf("cannot fetch block by number: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock by number (%v): %w", config.L1StartingBlockTag.BlockNumber, err)
 		}
 	}
 
-	// Ensure that there is a starting L1 block
-	if l1StartBlock == nil {
-		return fmt.Errorf("no starting L1 block")
+	// We want to ensure that we have sufficient time to start the sequencer
+	// before it exceeds max sequencer drift which is currently 2892s ==
+	// 48.2mins. Devops need about 7 mins to get the sequencer started after the
+	// migration script is run. We give them a 10 minute buffer in case of any
+	// issues, but if we are outside of this range we will fail here.
+	var maxSequencerDriftCelo uint64 = 2892 // Duplicate of the value defined at op-node/rollup.maxSequencerDriftCelo
+	if maxSequencerDriftCelo-(opts.migrationBlockTime-l1StartBlock.Time()) < 17*60 {
```

Comment:
If we'd hit this, this would break the migration. Shouldn't we instead try to find a newer finalized block?

Reply:
Yes, I think we can try to find a newer finalized block?

Reply:
Actually, if we assume that blocks explicitly set by tag are correct,  this check here is needless, because L1 block times are determined by the slot time, meaning that here the max diff between the l1 block time and l2 start time is (in the case that our L2 start time is the last second of an Eth 2 epoch):

`(32*12*3) + (12*9) - 1 = 1259s` ~= 20.98 minutes. We minus the one because the L2 start block is falling on the last second of the epoch in this scenario.

So given our maxSequencerDrift of 48.2m that gives us at least 27 minutes of time to get the sequencer up and running.

So this check would never get triggered with the current code.

Reply:
I've updated the logic to search the entire window between `l2StartBlockTime-maxSequencerDrift` and `l2StartBlockTime` for a finalized L1 block to use as the starting block.